tree: SharedTree Philosophy (#17201)
microsoft/FluidFramework: packages/dds/tree/docs/Shared Tree Philosophy.md

# Shared Tree Design Philosophy

> The goal of the shared tree project is empower developers to create and maintain better collaborative experiences.

Reviewer: This doc sometimes use "shared tree" and sometimes uses "Shared Tree". [This doc](https://fluidframework.com/docs/data-structures/tree/) uses `SharedTree`. Let's settle on one.

Author: Public API entry point is also SharedTree, so thats another reason to align with that (The other option would be to reference the package, but I don't think that works well here). I'll make the change.

Author: ```suggestion
The goal of the SharedTree project is empower developers to create and maintain better collaborative experiences.
```

    Users of shared tree can start by authoring their own specialized version of any of these (if needed), and then consider generalizing them as either independent libraries or upstream contributions.

Another way to describe these goals is that Shared Tree needs to be able to be **adopted with confidence**.
Once adopted, it must **compatible with existing systems**, **easily learnable** and **productive** for developers new to collaborative application while being **flexible** in what it supports and produce **robust** and **high quality** collaborative experiences.

In addition to these overall goals, Shared Tree also needs to provide an MVP early in the development and internally be **maintainable** and **extensible**.
This allows it to deliver value early, but also continue to be developed to reduce the development effort required of application authors to improve the experience it enables for end users.
These internal requirements mostly follow from the above flexibility requirement, but are worth the extra emphasis.

# Balancing Conflicting Goals

These goals are values which the Shared Tree tries to maximize.

> There are however cases where trad-offs must be made.

Reviewer: ```suggestion
There are however cases where trade-offs must be made.
```

Additionally the Shared Tree architecture organizes these components such that if needed they can be replaced incrementally.
For example, the tree reading and editing API is built on-top of cursors.
A different version of this API can be authored, tested and adopted side by side with the old one with minimum difficulty.
If the underlying tree storage (`forest`) changes, the tree API components (`flex-tree` (internal) and `simple-tree` (public)) will not be impacted.
Similarly if the cursor API needed to change, forest could add support for the new one (without dropping the old one), then users could migrate incrementally.
This kind of concurrent multi-version support is extra important for cases which fall into category one above (impact data compatibility not just API compatibility).
For example shared tree can introduce a new version of editing primitives and/or merge semantics for a field kind (for example sequence, see below),
which applications can opt into in their schema in a compatible way.

Shared Tree is also designed to ensure applications using it can also adopt this same compatibility approach while minimizing the frequency and difficulty of doing so.

> For example the APIs Shared Tree exposes for working with schema are designed to help guide users of it into design patterns that are robust and maintainable even when faced with supporting large numbers of legacy schema.

Reviewer: ```suggestion
For example the APIs Shared Tree exposes for working with schema are designed to help guide users into design patterns that are robust and maintainable even when faced with supporting large numbers of legacy schema.
```
or
```suggestion
For example the APIs Shared Tree exposes for working with schema are designed to help guide their usage into design patterns that are robust and maintainable even when faced with supporting large numbers of legacy schema.
```
or
```suggestion
For example the APIs Shared Tree exposes for working with schema are designed to help guide their users into design patterns that are robust and maintainable even when faced with supporting large numbers of legacy schema.
```

There are also a few kinds of fields which provide different collaborative data-structures (like `sequence`, or `optional`), each of which hold nodes and define what edits can be made to them.

All of the editing is done in the fields via the `field kind`.
This allows the editing logic, both API and merge policy, to be packaged together into minimal versionable units (`field kinds`).
Since the schema explicitly selects what field kinds to use, updating them is opt-in, and can be coordinated with deployment schedules to achieve the required cross version collaboration requirements.
This also leverages all the same collaboration as the rest of schema evolution, making adopting alternative field kinds no different from updating other aspects of the application's schema.

This results in a tree where the overall shape is controlled by the application via its tree schema (for the nodes) and the alternating layers of fields provide all the complex logic, neatly separating the concerns.
This can also be thought of as a tree of entities alternating between fields and nodes.

> As these two kinds of entities alternate and thus never touch, this resembles a [Bipartite Graph](https://en.wikipedia.org/wiki/Bipartite_graph), and can be called a Bipartite Tree.

Reviewer: Seems odd to capitalize these
```suggestion
As these two kinds of entities alternate and thus never touch, this resembles a [bipartite graph](https://en.wikipedia.org/wiki/Bipartite_graph), and can be called a bipartite tree.
```

Having this option available ensues the application author can always author their editing logic with confidence it won't behave unexpectedly due to concurrency, which is key to ensuring they can be productive.

This general purpose no-concurrency constraint is unfortunately not great for the "The merges apply both sets of edits" requirement.
This can be addressed by providing more specific constraints, and designing the editing APIs so constraints are needed in less cases.
For example allowing a constraint that only rejects a transaction if a field or subtree was concurrently edited, or providing a field kind with higher level editing operations like a set or counter with appropriate editing methods.

This design pattern ensures there is always a way for the application author to ensure correctness, and by a little doing more work to be less conservative with the constraints, the application can improve the merge granularity when needed.

Additionally, even higher merge fidelity can be achieved by adding new field kinds as application needs for them arise, and this work can be shared among all users of SharedTree.

> Note that the current implementation does not yet provide access to constraints, but this is planned for future versions.

Reviewer: Is that still true? I thought @noencke was adding that to the API recently.

Reviewer 2: They're not checked in, unfortunately. We reprioritized some other work instead for the upcoming deadlines.

Reviewer: Feel free to resolve without further changes

# Sub-Trees

To help applications scale to larger schema and more complex documents, SharedTree is designed to enable applications to process different subtrees independently.

> If the application aligns their data model with this, keeping invariants and operations as local to a given subtree as possible, then can benefit in several ways.

Reviewer: ```suggestion
If the application aligns their data model with this, keeping invariants and operations as local to a given subtree as possible, they can benefit in several ways.
```

It allows for easy to express but still fine grained constraints, using subtree constraint (which reject concurrent edits to the same subtree),
as well as more efficient and simple application logic for things like producing and updating a user interface for a given subtree's content.
The part of an application which handles a given subtree can be though of as a mini application which can be nested in a larger one:
this can both be great for reuse of logic, as well as integration into other tree structured systems, like [React](https://react.dev/) or even just the [DOM](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Introduction.)

> When operations need to span different subtrees, they can be though of as an operation at (or above) the level of their common parent, and can be constrained according when needed.

Reviewer: ```suggestion
When operations need to span different subtrees, they can be though of as an operation at (or above) the level of their common parent, and can be constrained accordingly when needed.
```

---

**NOTE**

This is part of why SHaredTree is a tree, and not a graph.
When dealing with mutable content, its really useful to be clear about what the invariants of any given data are.

> Having these comes from its schema is ideal, but in the cases where thats insufficient for constraints be contextually applied by the parents.

Reviewer: Some parts might be missing here. Here's an attempt:
```suggestion
Having these comes from its schema is ideal, but in the cases where that's insufficient, constraints can be contextually applied by the parents.
```

# Stored and View Schema

Applications need to make assumptions about the data they are working with.
For simple in-memory cases, this is generally managed with types.
The code in the application uses those types to enable readers of the code (including other developers and compilers) to reason about and validate the logic in the application.
Shared Tree provides this same experience: the application code gets to use types to reason about the document content it's working with, however it also has to deal with the fact that the document content might come from a different application or a different version of the same application (newer, older or even a divergent fork).
To ensure content from a document is not misinterpreted, the document carries with it the "stored schema", which defines the structure and editing behaviors of the content within the document.
This can be checked against the "view schema" which defines the types the application is programmed against.
If they match, the application can safely work on the document as is.
If they do not match, the application needs to take some corrective action.

> The set of possible corrective actions generally includes raising a compatibility error, updating or adapting the document to a supported format.

Reviewer: ```suggestion
The set of possible corrective actions generally includes raising a compatibility error, or updating or adapting the document to a supported format.
```
I think this helps clarify that it is the doc that is getting updated.

Author: The intention here is highlight the difference between the updating the schema chase and the case where the document content needs changes.
```suggestion
The set of possible corrective actions generally includes raising a compatibility error, updating the document's schema, or adapting the document's content to a supported format.
```
